// @webantic/util/lib/array.js
"use strict";
/** @module util/array */
exports.__esModule = true;
/**
 * @function move
 * @export
 * @summary moves an element at a given index to
 * another index
 * @param {Array<any>} array an array containing the
 * element to move
 * @param {number} fromIndex the location of the element
 * to move
 * @param {number} toIndex the destination index
 * @returns {Array<any>}
 */
function move(array, fromIndex, toIndex) {
    var copy = array.slice();
    var element = copy[fromIndex];
    copy.splice(fromIndex, 1);
    copy.splice(toIndex, 0, element);
    return copy;
}
exports.move = move;
/**
 * @function addToSet
 * @export
 * @summary Returns an array containing `item`, adding
 * it if it is missing.
 * Note: this only uses indexOf for comparison, so it
 * will not work for objects
 * @param {Array<any>} array an array to potentially add
 * `item` to
 * @param {any} item an item to ensure is within `array`
 * @returns {Array<any>} the new array
 */
function addToSet(array, item) {
    var copy = array.slice();
    if (copy.indexOf(item) === -1) {
        copy.push(item);
    }
    return copy;
}
exports.addToSet = addToSet;
/**
 * @function addToSetInPlace
 * @export
 * @summary Adds `item` to `set` if it is
 * missing, returns true if `set` was modified
 * @param {Array<any>} set a set to potentially add
 * `item` to
 * @param {any} item an item to ensure is within `set`
 * @returns {boolean} true if `set` was modified
 */
function addToSetInPlace(set, item) {
    if (set.indexOf(item) === -1) {
        set.push(item);
        return true;
    }
    return false;
}
exports.addToSetInPlace = addToSetInPlace;
/**
 * @function first
 * @export
 * @summary returns the first element in an array
 * @param {any[]} array the array to search in
 * @returns {any} the first element of the array
 */
function first(array) {
    if (array.length) {
        return array[0];
    }
}
exports.first = first;
/**
 * @function last
 * @export
 * @summary returns the last element in an array
 * @param {any[]} array the array to search in
 * @returns {any} the last element of the array
 */
function last(array) {
    if (array.length) {
        return array[array.length - 1];
    }
}
exports.last = last;
/**
 * @function eachUntil
 * @export
 * @summary calls `callback` for each item in an array,
 * provided that the previous iteration didn't return `false`
 * @param {any[]} array of items to iterate
 * @param {ForEachCallback} callback the function to invoke
 * for each element in the supplied array
 */
function eachUntil(array, callback) {
    for (var i = 0; i < array.length; i++) {
        if (callback(array[i], i, array) === false) {
            break;
        }
    }
}
exports.eachUntil = eachUntil;
/**
 * @function remove
 * @export
 * @summary Remove an item from an array
 * @param array {any[]} The array to remove `item` from
 * @param item {any} The item to remove. Cannot be an object
 * @returns {any[]} The new array
 * @example
 * var array = ['a', 'b', 'c']
 * remove(array, 'b') // returns ['a', 'c']
 */
function remove(array, item) {
    var clonedArray = Array.prototype.slice.call(array);
    clonedArray.splice(clonedArray.indexOf(item), 1);
    return clonedArray;
}
exports.remove = remove;
/**
 * @function removeAll
 * @export
 * @summary Removes all occurrences of `item` from `array`
 * @param {any[]} array The array to remove all occurrences of `item` from
 * @param {any} item The item to remove
 * @returns {any[]} The new array
 * @example
 * var array = ['a', 'b', 'b', 'c']
 * removeAll(array, 'b') // returns ['a', 'c']
 */
function removeAll(array, item) {
    var clonedArray = Array.prototype.slice.call(array);
    while (clonedArray.indexOf(item) !== -1) {
        clonedArray = remove(clonedArray, item);
    }
    return clonedArray;
}
exports.removeAll = removeAll;
/**
 * @function forceArray
 * @summary wraps the input in an array if it isn't already one
 * @param {any} input
 * @returns {any[]} always an array
 */
function forceArray(input) {
    return Array.isArray(input) ? input : [input];
}
exports.forceArray = forceArray;
